### PR TITLE
Rui/docker support

### PR DESCRIPTION
**Pull request by bbaayezi**
### Purpose

As mentioned in this issue #53 , this PR creates an option for the developer to create a docker container as dev environment, which eliminates the platform-related bugs during the setting up phase of the project.

### Changes

adding new `npm run start:dev` command and `start-dev-container.sh` script, the command will check for docker environment and setting up a development container for the user with hot module reload support. Dependencies are installed inside the container.

### New Requirements

docker environment is needed

### CODE REVIEW

**Comment by louislefevre** [comment]
@XxWirusonxX @Xu369 Could you two pull this branch and test the changes out please to see if it resolves your issues?

**Comment by XxWirusonxX** [comment]
Sure, im not familiar with Docker, is there any Installation Guide ? :) thanks @bbaayezi

**Comment by bbaayezi** [comment]
> Sure, im not familiar with Docker, is there any Installation Guide ? :) thanks @bbaayezi

You can check the official site [here](https://docs.docker.com/engine/install/).

**Review by louislefevre — CHANGES_REQUESTED** [review]
A few things:
- Currently, execution doesn't stop when an error occurs. Below are a few examples of commands I ran, followed by the errors I received. As you can see, the same errors are output multiple times due to the bash script not exiting when an error occurs.
- Could you write a short but detailed guide in the README on setting up the Docker environment, from complete scratch to a complete installation? We want to make it easy as possible for people to set up the environment, and to do that it needs to be documented.
- It would be good if you could add comments to the bash file to explain what each section of the code is doing.

`npm run start:dev`
```
docker installed
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Development container not running
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Creating volumes to store node dev dependencies
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Creating volumes to store python dev dependencies
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Image not found, building image and starting dev container
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

`systemctl start docker`
`npm run start:dev`
```
docker installed
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/containers/json?all=1": dial unix /var/run/docker.sock: connect: permission denied
Development container not running
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/volumes": dial unix /var/run/docker.sock: connect: permission denied
Creating volumes to store node dev dependencies
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/volumes/create": dial unix /var/run/docker.sock: connect: permission denied
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/volumes": dial unix /var/run/docker.sock: connect: permission denied
Creating volumes to store python dev dependencies
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/volumes/create": dial unix /var/run/docker.sock: connect: permission denied
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/images/json": dial unix /var/run/docker.sock: connect: permission denied
Image not found, building image and starting dev container
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/build?buildargs=%7B%7D&cachefrom=%5B%5D&cgroupparent=&cpuperiod=0&cpuquota=0&cpusetcpus=&cpusetmems=&cpushares=0&dockerfile=Dockerfile.dev&labels=%7B%7D&memory=0&memswap=0&networkmode=default&rm=1&shmsize=0&t=docker-quantum&target=&ulimits=null&version=1": dial unix /var/run/docker.sock: connect: permission denied
```

`sudo npm run start:dev`
```
docker installed
WARNING: Error loading config file: /root/.docker/config.json: open /root/.docker/config.json: permission denied
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/containers/json?all=1": dial unix /var/run/docker.sock: connect: permission denied
Development container not running
WARNING: Error loading config file: /root/.docker/config.json: open /root/.docker/config.json: permission denied
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/volumes": dial unix /var/run/docker.sock: connect: permission denied
Creating volumes to store node dev dependencies
WARNING: Error loading config file: /root/.docker/config.json: open /root/.docker/config.json: permission denied
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/volumes/create": dial unix /var/run/docker.sock: connect: permission denied
WARNING: Error loading config file: /root/.docker/config.json: open /root/.docker/config.json: permission denied
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/volumes": dial unix /var/run/docker.sock: connect: permission denied
Creating volumes to store python dev dependencies
WARNING: Error loading config file: /root/.docker/config.json: open /root/.docker/config.json: permission denied
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/volumes/create": dial unix /var/run/docker.sock: connect: permission denied
WARNING: Error loading config file: /root/.docker/config.json: open /root/.docker/config.json: permission denied
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/images/json": dial unix /var/run/docker.sock: connect: permission denied
Image not found, building image and starting dev container
WARNING: Error loading config file: /root/.docker/config.json: open /root/.docker/config.json: permission denied
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/build?buildargs=%7B%7D&cachefrom=%5B%5D&cgroupparent=&cpuperiod=0&cpuquota=0&cpusetcpus=&cpusetmems=&cpushares=0&dockerfile=Dockerfile.dev&labels=%7B%7D&memory=0&memswap=0&networkmode=default&rm=1&shmsize=0&t=docker-quantum&target=&ulimits=null&version=1": dial unix /var/run/docker.sock: connect: permission denied

```

**Comment by XxWirusonxX** [comment]
Was running out of space to install Docker yesterday lmao, im going to test it.. or should i wait until next Push if Louis has done that already?

**Comment by Xu369** [comment]
> @XxWirusonxX @Xu369 Could you two pull this branch and test the changes out please to see if it resolves your issues?

yeah, sure, though my issues have been resolved, I will test it to see if it works

**Comment by louislefevre** [comment]
Going to close this PR for now. We can reopen it once we return to working on this.

**Review by KhesimReid — APPROVED** [review]
Amazing work Rui! It works really well and should mitigate against any machine dependent issues we've yet to encounter and fix.

**Review by ttoktassynov — APPROVED** [review]
LGTM, good job!

**Review by louislefevre — CHANGES_REQUESTED** [review]
```
Step 4/8 : COPY package*.json .
When using COPY with more than one source file, the destination must be a directory and end with a /
ERROR: Service 'web' failed to build : Build failed
```
Step 4 fails for me. I think this is due to changes in later versions of Docker. My versions are:
- Docker 20.10.8
- Docker-compose 1.29.2

**Comment by bbaayezi** [comment]
> ```
> Step 4/8 : COPY package*.json .
> When using COPY with more than one source file, the destination must be a directory and end with a /
> ERROR: Service 'web' failed to build : Build failed
> ```
> 
> Step 4 fails for me. I think this is due to changes in later versions of Docker. My versions are:
> 
> * Docker 20.10.8
> * Docker-compose 1.29.2

@louislefevre can you test if change `COPY package*.json .` to `COPY package*.json ./` solves the issue?

File: Dockerfile.dev [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+RUN npm install -g nodemon
+
+WORKDIR /node-red-contrib-quantum
+
+COPY package*.json .
+
+COPY bin ./bin
+
+RUN npm install && \
+    npm run setup
```
Comment (louislefevre) [inline Dockerfile.dev:13]:
```suggestion
RUN npm run setup
```
This can be simplified to just `npm run setup`, since that command runs install anyway.

**Comment by louislefevre** [comment]
> @louislefevre can you test if change `COPY package*.json .` to `COPY package*.json ./` solves the issue?

Yup, that seems to fix it 👍

**Comment by louislefevre** [comment]
Also, where are the files created for the Docker environment stored? So the virtual environment, `.node-red` directory, etc. Where can that be found?

**Comment by bbaayezi** [comment]
> Also, where are the files created for the Docker environment stored? So the virtual environment, `.node-red` directory, etc. Where can that be found?

@louislefevre This line in `docker-compose.yml` creates a docker volume and bind it with the `.node-red` directiory:
https://github.com/node-red-quantum/node-red-contrib-quantum/blob/6cfa28ba3b2d2cb85b797e70365d1932e0afd83b/docker-compose.yml#L17

so when you run `docker volume ls`, there should be a volume named `quantum_flows`, that's where the flow configs are stored. The content inside a docker volume can be transferred to the host machine or to other containers, but currently, there is no explicit tutorial in README showing how to achieve that, maybe we can add it in the future PR.

**Review by louislefevre — APPROVED** [review]

**Pull request merged**